apps/src/templates/progress/ProgressBubble.jsx

```diff
@@ -35,272 +34,78 @@ class ProgressBubble extends React.Component {
       PropTypes.string,
       PropTypes.number
     ]),
-    // This prop is provided as a testing hook, in normal use it will just be
-    // set to window.location; see defaultProps.
-    currentLocation: PropTypes.object.isRequired,
-    lessonTrophyEnabled: PropTypes.bool,
-    pairingIconEnabled: PropTypes.bool,
     hideToolTips: PropTypes.bool,
-    hideAssessmentIcon: PropTypes.bool,
     onClick: PropTypes.func
   };
 
-  static defaultProps = {
-    currentLocation: window.location
-  };
-
-  recordProgressTabProgressBubbleClick = () => {
-    firehoseClient.putRecord(
-      {
-        study: 'teacher_dashboard_actions',
-        study_group: 'progress',
-        event: 'go_to_level',
-        data_json: JSON.stringify({
-          student_id: this.props.selectedStudentId,
-          level_url: this.props.level.url,
-          section_id: this.props.selectedSectionId
-        })
-      },
-      {includeUserId: true}
-    );
-  };
-
-  render() {
-    const {
-      level,
-      disabled,
-      smallBubble,
-      selectedSectionId,
-      selectedStudentId,
-      currentLocation,
-      lessonTrophyEnabled,
-      pairingIconEnabled,
-      hideAssessmentIcon,
-      onClick
-    } = this.props;
-
-    const levelIsAssessment = isLevelAssessment(level);
-
-    const number = level.levelNumber;
-    const url = level.url;
-    const levelName = level.name || level.progressionDisplayName;
-    const levelIcon = getIconForLevel(level);
-
-    const hideNumber =
-      level.letter || level.isLocked || level.paired || level.bonus;
-
-    const style = {
-      ...styles.main,
-      ...(!disabled && hoverStyle),
-      ...(smallBubble && styles.small),
-      ...(level.isConceptLevel &&
-        (smallBubble ? styles.smallDiamond : styles.largeDiamond)),
-      ...levelProgressStyle(level.status, level.kind),
-      ...(level.highlighted && styles.highlighted)
-    };
-
-    let href = '';
-    if (!disabled && url && !onClick) {
-      const queryParams = queryString.parse(currentLocation.search);
+  isEnabled() {
+    const {disabled, level, onClick} = this.props;
+    return !disabled && (!!level.url || !!onClick);
+  }
 
-      if (selectedSectionId) {
-        queryParams.section_id = selectedSectionId;
-      }
-      if (selectedStudentId) {
-        queryParams.user_id = selectedStudentId;
-      }
-      const paramString = queryString.stringify(queryParams);
-      href = url;
-      if (paramString.length > 0) {
-        // If href already has 1 or more query params, our delimiter will be '&'.
-        // If href has no query params, our delimiter is '?'.
-        // TODO: (madelynkasula) Refactor this logic to use queryString.parseUrl(href)
-        // instead. Our current version of query-string (4.1.0) does not yet have this method.
-        const delimiter = /\?/.test(href) ? '&' : '?';
-        href += delimiter + paramString;
-      }
+  onClickLevel = () => {
```

Comment:
For my education, why is this function an arrow function but `getUrl` and `createBubbleElement` are not?

Reply:
great edifying question! `getUrl` and `createBubbleElement` are called from `render`, which is already bound to `this`.  `onClickLevel` is passed into `BubbleLink` as a callback, and since it references `this.props`, that means it needs to be explicitly bound to `this`, and arrow functions are auto-bound to the context in which they're declared (in this case, on the instance of the class). if i had made this a regular function, i would have needed to add a constructor and manually bound the function there.

Reply:
Thanks for the explanation!  Understanding _this_ requires too much mental gymnastics (pun intended).  :-)